Add/better error message for minimum required amount

dmvrtx opened the pull request:
Fixes #701 

#### Changes proposed in this Pull Request

* Localized error message if order amount is too low to process, based on payments server response.

#### Testing instructions

 * On a connected WooCommerce store create a product with a price below 0.5 USD
 * Put it into cart and try to purchase through WCPay
 * Error message should say:`Sorry, the minimum allowed order total is $0.50 to use this payment method`
  
<img width="800" alt="image" src="https://user-images.githubusercontent.com/683297/106288936-f30fd300-6248-11eb-8f13-b3ed873b9d05.png">
-------------------

- [x] Added changelog entry (or does not apply)
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

<!--
Please read P7bbVw-3ww-p2 before opening the PR, assign the correct status to it, __and make sure that the related issue uses the Has PR status!__.
-->

dmvrtx committed:
Provide a better error message for minimum order amount requirement

WooCommerce payments might return an error on payment intent creation
if order amount is too low. This error will also contain a minimum order
amount value alongside with currency so client can present it in its own
way.

This change makes WooCommerce plugin to use this information to provide
a localized message to the end user.

dmvrtx committed:
Update changelog and readme

dmvrtx committed:
Merge branch 'trunk' into add/better-error-message-for-minimum-required-amount

dmvrtx committed:
Render expected message instead of hard-coding to make tests more robust

dmvrtx committed:
fix php linter error (no comma after the last function arg)

kalessil reviewed (changes requested):
Overall looks great, just a few comments left.

dmvrtx committed:
Address review notes

dmvrtx commented on includes/wc-payment-api/class-wc-payments-api-client.php:
```diff
@@ -991,6 +991,29 @@ protected function extract_response_body( $response ) {
+
+				// Response contains a minumum order amount which we will
+				// use to build a specific message for a client.
+				$message = sprintf(
+					// translators: %1$s is a formatted amount with currency code.
+					__(
+						'Sorry, the minimum allowed order total is %1$s to use this payment method.',
+						'woocommerce-payments'
+					),
+					wc_price( $min_amount, [ 'currency' => strtoupper( $currency ) ] )
```
@marcinbot, my main reason to parse it here was that I pass some additional data from the server side, i.e. minimum required amount and (settlement) currency. Existing flows seem to discard that additional data.

On the other hand I agree, it might be better to extend `API_Exception` and make it carry additional data along. I'll take a look at it.

jessepearson replied:
Added that mentioned PR here: https://github.com/Automattic/woocommerce-payments/pull/1268

dmvrtx replied:
I decided to go a similar way, but to make an exception class to return a properly formatted message instead of replacing it when notice is created.

dmvrtx committed:
Apply code reviews suggestions:

 * Error message formatting is done through the
 `woocommerce_exception_message` filter which gets error message and
 exception object.
 * `WCPay\Exceptions\API_Exception` now can carry `additional_data`
 as a machine-readable array.
 * Exception handling is moved to the gateway class rather than API
 client.

dmvrtx committed:
Merge branch 'trunk' into add/better-error-message-for-minimum-required-amount

dmvrtx committed:
Fix after-merge whitespace issues

dmvrtx committed:
Merge branch 'trunk' into add/better-error-message-for-minimum-required-amount

dmvrtx committed:
Fix a subtle merge issue

dmvrtx committed:
Merge branch 'trunk' into add/better-error-message-for-minimum-required-amount

dmvrtx committed:
Whitespace fix in the tests

dmvrtx committed:
Merge branch 'trunk' into add/better-error-message-for-minimum-required-amount

dmvrtx committed:
Re-arrange API_Exception arguments to match convention

dmvrtx commented on includes/exceptions/class-api-exception.php:
```diff
@@ -20,17 +20,26 @@ class API_Exception extends Base_Exception {
 	 */
-	public function __construct( $message, $error_code, $http_code, $code = 0, $previous = null ) {
-		$this->http_code = $http_code;
+	public function __construct( $message, $error_code, $http_code, $code = 0, $previous = null, $additional_data = null ) {
```
Good point. Done.

dmvrtx commented on includes/class-wc-payments-utils.php:
```diff
@@ -355,4 +355,62 @@ public static function get_order_intent_currency( WC_Order $order ): string {
+	 *
+	 * @return string User-friendly error message.
+	 */
+	public static function wc_format_amount_too_low_exception( $e ) {
```
I'd actually go for a new exception type then and override `getMessage()` to format it accordingly. But this might require to introduce a factory method to create a proper exception based on error code.

dmvrtx committed:
Remove helper formatting functions in favour of Exception object method

dmvrtx committed:
Move conditional logic into sub-class

dmvrtx commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -470,7 +470,7 @@ public function process_payment( $order_id ) {
 			if ( is_a( $e, Connection_Exception::class ) ) {
 				$error_message = __( 'There was an error while processing the payment. If you continue to see this notice, please contact the admin.', 'woocommerce-payments' );
 			}
```
I suggest move it into the `Connection_Exception` instead if it makes sense for everyone. Similar to how it is done in `Amount_Too_Low_Exception`.

dmvrtx commented on includes/class-wc-payments-utils.php:
```diff
@@ -355,4 +355,62 @@ public static function get_order_intent_currency( WC_Order $order ): string {
+	 *
+	 * @return string
+	 */
+	public static function wc_format_api_exception_message( $message, $e ) {
```
Removed them completely in favour of formatting a message in the exception class.

vbelolapotkov reviewed (commented):
@dmvrtx I haven't tested it properly but code looks good. Added a few comments.

vbelolapotkov commented on tests/unit/exceptions/test-class-wc-payments-amount-too-low-exception.php:
```diff
@@ -0,0 +1,52 @@
+					'Sorry, the minimum allowed order total is %1$s to use this payment method.',
+					'woocommerce-payments'
+				),
+				wc_price( $error_data['min_amount'] / 100, [ 'currency' => strtoupper( $error_data['currency'] ) ] )
```
Since this is known value in test, we might even use `$0.60` inside message without sprintf :)

dmvrtx replied:
`wc_price` does some wrapping with `<span>` elements and it might differ from version to version, hence I use it in the tests to match generated message in the exception.

vbelolapotkov replied:
That's even bigger question, do we need this span in error message and anywhere else?

dmvrtx replied:
Yeah, an interesting question indeed. If we consider `API_Exception` and its derivatives as something that can be directly communicated to the user, then I see no issue with `<span>` as long as it is properly documented. Other use cases can strip HTML tags if needed.

vbelolapotkov replied:
I'm still look at this addition of span as a side effect. I wasn't aware of it and didn't expect it until you've mentioned the span is added. Addressing this globally might be out of scope of this PR though. Anyway, I'd prefer us using literal strings for setting expectations rather than something calculated with the same utils we use in the code under test.

dmvrtx replied:
Addressed in the related change to price formatting method.

dmvrtx committed:
Use existing amount adjustment function

dmvrtx commented on includes/exceptions/class-amount-too-low-exception.php:
```diff
@@ -0,0 +1,54 @@
+
+			if ( ! in_array( $currency, \WC_Payments_Utils::zero_decimal_currencies(), true ) ) {
+				$min_amount = $min_amount / 100;
+			}
```
Re-used `interpret_stripe_amount`, but I am not sure if I understand the part about the `wc_price`.

vbelolapotkov replied:
> Re-used interpret_stripe_amount
👍 

>  I am not sure if I understand the part about the wc_price.

Sorry, for the vague comment. I was thinking about creating utility method like `\WC_Payments_Utils::format_currency( $amount, $currency )` which would check if currency is zero decimal and adjust amount as needed. Then in this error handler we could use it right inside sprintf like this: `\WC_Payments_Utils::format_currency( $additional_data['min_amount'], additional_data['currency'] )`. 

This would allow us to move away from using `wc_price` directly everywhere in the code base and with time replace `wc_price` with own implementation at all (to solve <span> problem). 

What do you think?

dmvrtx replied:
> Sorry, for the vague comment. I was thinking about creating utility method like \WC_Payments_Utils::format_currency( $amount, $currency ) 

I like this approach. It would also address our `<span>` discussion. As a first naive implementation I would just call `wc_price` wrapped with `strip_tags` to remove any HTML formatting.

dmvrtx replied:
I am still on the fence with the currency symbol. `wc_price` will try to use an HTML-entity for a currency symbol and my current implementation of `format_amount` should convert it into a (Unicode) character (depends on a symbol). Which may play nicely with how it is later used in the notice but may be not.

dmvrtx committed:
Move variable declaration closer to the usage

dmvrtx committed:
Merge branch 'trunk' into add/better-error-message-for-minimum-required-amount

dmvrtx commented on tests/unit/exceptions/test-class-wc-payments-amount-too-low-exception.php:
```diff
@@ -0,0 +1,52 @@
+		$e_with_data = new WCPay\Exceptions\Amount_Too_Low_Exception( 'Test', 'wcpay_amount_too_low', 400, $error_data );
+		$this->assertEquals(
+			sprintf(
+				// translators: %1$s is a formatted amount with currency code.
+				__(
```
I kept it there for the cases if we start running tests in multiple languages for any reason.

vbelolapotkov replied:
I doubt we'll test translations this way because AFAIK translations are handled via WP outside of the plugin. Anyway, this is a minor thing, so leaving decision up to you.

dmvrtx replied:
True, removed translations in tests for now.

dmvrtx committed:
Make exception message text-only.

`wc_price` tends to bloat resulting string with HTML tags and entities,
so to use it inside an exception message and other non-HTML places it
has to be stripped down.

To simplify the process `format_amount` method is added to
`WC_Payment_Utils` to apply proper amount transformations for amounts
used with Stripe and to return a properly formatted price without tags
and entities.

dmvrtx committed:
Merge branch 'trunk' into add/better-error-message-for-minimum-required-amount

dmvrtx committed:
Move Connection_Exception message override into the exception.

dmvrtx commented:
Closing the PR as changes do not make a lot of sense anymore.
